src/xrpld/app/tx/detail/InvariantCheck.cpp

```diff
@@ -2170,4 +2179,911 @@ ValidAMM::finalize(
+                afterVault_->shareMPTID != beforeVault_->shareMPTID)
+            {
+                JLOG(j.fatal())
+                    << "Invariant failed: violation of vault immutable data";
+                result = false;
+            }
+        }
+
+        if (!updatedShares)
+        {
+            JLOG(j.fatal())
+                << "Invariant failed: updated vault must have shares";
+            XRPL_ASSERT(
+                enforce,
+                "ripple::ValidVault::finalize : vault has shares invariant");
+            return !enforce;  // That's all we can do here
+        }
+
+        if (updatedShares->sharesTotal == 0)
+        {
+            if (afterVault_->assetsTotal != zero)
+            {
+                JLOG(j.fatal()) << "Invariant failed: updated zero sized "
+                                   "vault must have no assets outstanding";
+                result = false;
+            }
+            if (afterVault_->assetsAvailable != zero)
+            {
+                JLOG(j.fatal()) << "Invariant failed: updated zero sized "
+                                   "vault must have no assets available";
+                result = false;
+            }
+        }
+        else if (updatedShares->sharesTotal > updatedShares->sharesMaximum)
+        {
+            JLOG(j.fatal())  //
+                << "Invariant failed: updated shares must not exceed maximum "
+                << updatedShares->sharesMaximum;
+            result = false;
+        }
+
+        if (afterVault_->assetsAvailable < zero)
+        {
+            JLOG(j.fatal())
+                << "Invariant failed: assets available must be positive";
+            result = false;
+        }
+
+        if (afterVault_->assetsAvailable > afterVault_->assetsTotal)
+        {
+            JLOG(j.fatal()) << "Invariant failed: assets available must "
+                               "not be greater than assets outstanding";
+            result = false;
+        }
+        else if (
+            afterVault_->lossUnrealized >
+            afterVault_->assetsTotal - afterVault_->assetsAvailable)
+        {
+            JLOG(j.fatal())  //
+                << "Invariant failed: loss unrealized must not exceed "
+                   "the difference between assets outstanding and available";
+            result = false;
+        }
+
+        if (afterVault_->assetsTotal < zero)
+        {
+            JLOG(j.fatal())
+                << "Invariant failed: assets outstanding must be positive";
+            result = false;
+        }
+
+        if (afterVault_->assetsMaximum < zero)
+        {
+            JLOG(j.fatal())
+                << "Invariant failed: assets maximum must be positive";
+            result = false;
+        }
+    }
+
+    if (txnType == ttVAULT_CREATE ||    //
+        txnType == ttVAULT_SET ||       //
+        txnType == ttVAULT_DEPOSIT ||   //
+        txnType == ttVAULT_WITHDRAW ||  //
+        txnType == ttVAULT_CLAWBACK)
```

Comment:
This should probably be a "permission"

Reply:
Thanks, added `mustModifyVault` (and a TODO to add `mayModifyVault` which I think you will do)